### ISSUE TITLE
Unable to create save directory (or write anything in general) if its path contains a symlink

### ISSUE THREAD

**Issue by Svalorzen** [love2d/love#1880]
I just spent a bunch of hours debugging the issue I reported [here](https://love2d.org/forums/viewtopic.php?f=4&t=94044) on the LOVE forums.

The jist is this: inside the physfs library, which LOVE bundles, the function that is supposed to create directories is called `doMkdir`. This function recursively traverses the input path and creates directories as it goes along. However, it performs a check on the parts that already exist, forcing them to be directories themselves. If that's not the case, it stops. The check is [the following](https://github.com/love2d/love/blob/main/src/libraries/physfs/physfs.c#L2190):

```
retval = ((rc) && (statbuf.filetype == PHYSFS_FILETYPE_DIRECTORY));
```

However, this breaks if the save path contains a symlink. In this case, the filetype will be a symlink and not a directory, the makedir will fail, LOVE will fail to set its main writing directory, and no writing will be possible. This is the case for me, as my `.local` folder is a symlink to a folder stored on an SSD, different to where I keep my home folder.

It seems that the current version of physfs has this patched: [this line](https://github.com/icculus/physfs/blob/main/src/physfs.c#L2242) now reads:

```
retval = ( (rc) && ((statbuf.filetype == PHYSFS_FILETYPE_DIRECTORY) || (statbuf.filetype == PHYSFS_FILETYPE_SYMLINK)) );
```
which correctly addresses the problem.

I did not check exactly the differences between the current version of physfs and the one embedded by LOVE, so I'm unsure if there's any other bugfixes or API changes. However, it might make sense to consider updating it; in my case just updating the single line I mentioned was enough to fix the problem (I cloned and built love myself so I could debug this issue).

**Comment by slime73** [comment]
While I want to update love 12's physfs regardless, can you try with love 12 and see whether it works? I don't think it calls PHYSFS_setWriteDir anymore, so it likely takes a different path which might be fixed already (or might need a separate fix if it doesn't work).

**Comment by Svalorzen** [comment]
I can try; as a question, is there a way to have a separate build folder for love or do I have to clean and rebuild every time? I'm using the configure/make build steps to compile.

**Comment by MikuAuahDark** [comment]
You can always out-of-tree build.

With CMake, it's
```sh
cmake -Bbuild -S.
```

With Autotools, it's
```sh
mkdir build
cd build
../configure
make -j
```

**Comment by Svalorzen** [comment]
I get an error during building 12.0-development, want me to report it here or should I open a separate issue?

**Comment by MikuAuahDark** [comment]
What I'm saying is both build system supports out-of-tree build.

I suggest sending the build issue in Discord, but if you can't, or won't use Discord, then opening separate issue is fine.

**Comment by Svalorzen** [comment]
Ok, managed to compile it. It seems that the issue is indeed fixed in 12.0-development! But so you are just making sure you don't use physfs to create any sort of directory anywhere in the code?

**Comment by slime73** [comment]
> you are just making sure you don't use physfs to create any sort of directory anywhere in the code?

No, `love.filesystem.createDirectory` still uses PhysFS. But the path you pass to it will always be inside an already created and mounted save directory, so it won't run into your issue because it won't have your `.local` symlink in the relative path you pass it.
